File: src/hosts/MacBook/scripts/generate-automator-thumbnails.m
// generate-automator-thumbnails.m — Render SF Symbol to 256×256 PNG via AppKit.
//
// Usage: generate-automator-thumbnails <symbol-name> <output-path>
//
// Compiles with: $CC -fobjc-arc -fmodules -framework AppKit -framework Foundation
//
// WHY Objective-C instead of Swift: pkgs.swift 5.10.1 is incompatible with
// nixpkgs apple-sdk 14.4 — Foundation module references CoreFoundation.CGFloat
// which Nix doesn't export. ObjC + clang from stdenv works correctly.

#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc != 3) {
            fprintf(stderr, "Usage: %s <symbol-name> <output-path>\n", argv[0]);
            return 1;
        }

        NSString *symbolName = [NSString stringWithUTF8String:argv[1]];
        NSString *outputPath = [NSString stringWithUTF8String:argv[2]];

        NSImage *image = [NSImage imageWithSystemSymbolName:symbolName
                                  accessibilityDescription:nil];
        if (!image) {
            fprintf(stderr, "Error: Could not create image for '%s'\n",
                    [symbolName UTF8String]);
            return 1;
        }

        NSImageSymbolConfiguration *config = [NSImageSymbolConfiguration
            configurationWithPointSize:120 weight:NSFontWeightMedium scale:NSImageSymbolScaleMedium];
        NSImage *configuredImage = [image imageWithSymbolConfiguration:config];
        if (!configuredImage) {
            fprintf(stderr, "Error: Could not configure image\n");
            return 1;
        }

        NSSize size = NSMakeSize(256, 256);
        NSBitmapImageRep *bitmapRep = [[NSBitmapImageRep alloc]
            initWithBitmapDataPlanes:NULL
            pixelsWide:size.width
            pixelsHigh:size.height
            bitsPerSample:8
            samplesPerPixel:4
            hasAlpha:YES
            isPlanar:NO
            colorSpaceName:NSCalibratedRGBColorSpace
            bytesPerRow:0
            bitsPerPixel:0];
        if (!bitmapRep) {
            fprintf(stderr, "Error: Could not create bitmap\n");
            return 1;
        }

        [NSGraphicsContext
            setCurrentContext:[NSGraphicsContext
                graphicsContextWithBitmapImageRep:bitmapRep]];
        [[NSColor whiteColor] set];
        NSRectFill(NSMakeRect(0, 0, size.width, size.height));
        [configuredImage drawInRect:NSMakeRect(0, 0, size.width, size.height)
                           fromRect:NSZeroRect
                          operation:NSCompositingOperationSourceOver
                           fraction:1.0];
        [NSGraphicsContext setCurrentContext:nil];

        NSData *data = [bitmapRep
            representationUsingType:NSBitmapImageFileTypePNG
                         properties:@{}];
        if (!data) {
            fprintf(stderr, "Error: Could not create PNG data\n");
            return 1;
        }

        NSError *error = nil;
        BOOL ok = [data writeToFile:outputPath options:NSDataWritingAtomic error:&error];
        if (!ok) {
            fprintf(stderr, "Error: Could not write to '%s': %s\n",
                    [outputPath UTF8String],
                    [[error localizedDescription] UTF8String]);
            return 1;
        }

        fprintf(stdout, "OK: %s (%lu bytes)\n",
                [outputPath UTF8String], (unsigned long)data.length);
    }
    return 0;
}

File: src/hosts/MacBook/scripts/generate-automator-thumbnails.swift
// generate-automator-thumbnails.swift — Generate QuickLook/Thumbnail.png for
// Automator workflow bundles from SF Symbol names.
//
// Usage: swift generate-automator-thumbnails.swift <symbol-name> <output-path>
//
// Renders a 256×256 PNG of the given SF Symbol at 120pt medium weight.
// Skips generation if the output file already exists (idempotent).
//
// WHY: Automator workflow bundles use QuickLook/Thumbnail.png for the Get Info
// window icon. Without it, macOS shows a generic workflow icon instead of the
// SF Symbol render. This script is run once per workflow to populate the PNG,
// which is then committed as a static asset.
//
// Requirements: macOS 11+ (SF Symbols), Swift with AppKit.

import AppKit
import Foundation

guard CommandLine.arguments.count == 3 else {
    fputs("Usage: \(CommandLine.arguments[0]) <symbol-name> <output-path>\n", stderr)
    exit(1)
}

let symbolName = CommandLine.arguments[1]
let outputPath = CommandLine.arguments[2]

// Skip if output already exists (idempotent re-runs).
if FileManager.default.fileExists(atPath: outputPath) {
    print("Skipping \(outputPath) — already exists")
    exit(0)
}

guard let image = NSImage(systemSymbolName: symbolName, accessibilityDescription: nil) else {
    fputs("Error: SF Symbol '\(symbolName)' not found\n", stderr)
    exit(1)
}

let config = NSImage.SymbolConfiguration(pointSize: 120, weight: .medium)
let configured = image.withSymbolConfiguration(config)!

let size = NSSize(width: 256, height: 256)
let bitmap = NSBitmapImageRep(
    bitmapDataPlanes: nil,
    pixelsWide: Int(size.width),
    pixelsHigh: Int(size.height),
    bitsPerSample: 8,
    samplesPerPixel: 4,
    hasAlpha: true,
    isPlanar: false,
    colorSpaceName: .deviceRGB,
    bytesPerRow: 0,
    bitsPerPixel: 0
)!

NSGraphicsContext.saveGraphicsState()
NSGraphicsContext.current = NSGraphicsContext(bitmapImageRep: bitmap)
NSGraphicsContext.current?.cgContext.scaleBy(x: 1, y: -1)
configured.draw(in: NSRect(origin: .zero, size: size), from: .zero, operation: .sourceOver, fraction: 1.0)
NSGraphicsContext.restoreGraphicsState()

guard let pngData = bitmap.representation(using: .png, properties: [:]) else {
    fputs("Error: failed to create PNG data\n", stderr)
    exit(1)
}

// Create parent directory if needed.
let dir = (outputPath as NSString).deletingLastPathComponent
if !FileManager.default.fileExists(atPath: dir) {
    try FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)
}

try pngData.write(to: URL(fileURLWithPath: outputPath))
print("Generated \(outputPath)")
